xds/src/main/java/io/grpc/xds/internal/rbac/engine/AuthorizationEngine.java

```diff
@@ -0,0 +1,208 @@
+    private final Action action;
+    private final ImmutableMap<String, Expr> conditions;
+
+    public RbacEngine(Action action, ImmutableMap<String, Expr> conditions) {
+      this.action = Preconditions.checkNotNull(action);
+      this.conditions = Preconditions.checkNotNull(conditions);
+    }
+  }
+
+  private final RbacEngine allowEngine;
+  private final RbacEngine denyEngine;
+
+  /**
+   * Creates a CEL-based Authorization Engine from one Envoy RBAC.
+   * @param rbacPolicy input Envoy RBAC policy.
+   */
+  public AuthorizationEngine(RBAC rbacPolicy) {
+    Map<String, Expr> conditions = new LinkedHashMap<>();
+    for (Map.Entry<String, Policy> policy: rbacPolicy.getPolicies().entrySet()) {
+      conditions.put(policy.getKey(), policy.getValue().getCondition());
+    }
+    allowEngine = (rbacPolicy.getAction() == Action.ALLOW) 
+        ? new RbacEngine(Action.ALLOW, ImmutableMap.copyOf(conditions)) : null; 
+    denyEngine = (rbacPolicy.getAction() == Action.DENY) 
+        ? new RbacEngine(Action.DENY, ImmutableMap.copyOf(conditions)) : null; 
+  }
+
+  /**
+   * Creates a CEL-based Authorization Engine from two Envoy RBACs.
+   * When it takes two RBACs, the order has to be a RBAC with DENY action
+   * followed by a RBAC with ALLOW action.
+   * @param denyPolicy input Envoy RBAC policy with DENY action.
+   * @param allowPolicy input Envoy RBAC policy with ALLOW action.
+   * @throws IllegalArgumentException if the user inputs an invalid RBAC list.
+   */
+  public AuthorizationEngine(RBAC denyPolicy, RBAC allowPolicy) throws IllegalArgumentException {
+    if (denyPolicy.getAction() != Action.DENY || allowPolicy.getAction() != Action.ALLOW) {
+      throw new IllegalArgumentException( "Invalid RBAC list, " 
+          + "must provide a RBAC with DENY action followed by a RBAC with ALLOW action. ");
+    }
+    Map<String, Expr> denyConditions = new LinkedHashMap<>();
+    for (Map.Entry<String, Policy> policy: denyPolicy.getPolicies().entrySet()) {
+      denyConditions.put(policy.getKey(), policy.getValue().getCondition());
+    }
+    denyEngine = new RbacEngine(Action.DENY, ImmutableMap.copyOf(denyConditions));
+    Map<String, Expr> allowConditions = new LinkedHashMap<>();
+    for (Map.Entry<String, Policy> policy: allowPolicy.getPolicies().entrySet()) {
+      allowConditions.put(policy.getKey(), policy.getValue().getCondition());
+    }
+    allowEngine = new RbacEngine(Action.ALLOW, ImmutableMap.copyOf(allowConditions));   
+  }
+
+  /**
+   * The evaluate function performs the core authorization mechanism of CEL Evaluation Engine.
+   * Determines whether a gRPC call is allowed, denied, or unable to decide.
+   * @param args evaluate argument that is used to evaluate the RBAC conditions.
+   * @return an AuthorizationDecision generated by CEL Evaluation Engine.
+   */
+  public AuthorizationDecision evaluate(EvaluateArgs args) {
+    List<String> unknownPolicyNames = new ArrayList<>();
```

Comment:
`unknownPolicyNames` belongs to this method and you are passing it to `evaluateEngine` twice and having that method return `UNKNOWN` based on the length of the array. What will be simpler/more elegant is to make the check in this method and have evaluateEngine return only ALLOW/DENY or null. Just a suggestion

Reply:
`unknownPolicyNames` is updated in `evaluateEngine` when going over each condition, so I have to pass it to `evaluateEngine`. If that is the case, shall I still need to move the check for `unknownPolicyNames` to `evaluate`?

Reply:
Up to you. If it were me I would move the check for `unknownPolicyNames` being non-empty in this method to enforce the logic you have described in the comments below. But you can leave this unchanged if you prefer.

Reply:
I'll change it then since that seems to make the logic more clear :)